### ISSUE TITLE
FLOATING-POINT-INVALID-OPERATION under SBCL, but not CCL 

### ISSUE THREAD

**Issue by seigakaku** [bohonghuang/cl-gtk4#30]
I've been working through multiple of your examples, including those in issues #25 and #19 and the GTKMM C++ demos of the new List API, with the ultimate goal of displaying a tree view with GtkTreeListModel and GtkColumView, however, the CFFI callback seems to cause an internal floating point error with SBCL, but not in CCL.

```common-lisp
(in-package #:gtk-test)

(proclaim '(optimize (safety 0) (speed 0)))

(cffi:defcallback tree-list-model-create-model-func :pointer ((item :pointer) (user-data :pointer))
  (declare (ignore user-data))
  (when item
    (let* ((uuid (string-object-string (gobj:pointer-object item 'string-object))))
      (print uuid)))
  (cffi:null-pointer))

(defstruct product-info
  name cost sale)

(define-application (:name closting :id "com.closting")
  (define-main-window (window (make-application-window :application *application*))


    (let* ((product-table (make-hash-table :test #'equal))
           (rows (loop :for (name cost sale) :in '(("HP Zbook 15" "250$" "300$")
                                                   ("Infinix S90" "40$" "60$")
                                                   ("Casio Calculator" "50$" "80$"))
                       :for product-info := (make-product-info :name name :cost cost :sale sale)
                       :for uuid := (prin1-to-string (uuid:make-v4-uuid))
                       :do (setf (gethash uuid product-table) product-info)
                       :collect uuid))
           (root-model (make-string-list :strings rows))
           (tree-list-model (make-tree-list-model :root (make-multi-selection :model root-model)
                                                  :passthrough nil
                                                  :autoexpand t
                                                  :create-func (cffi:callback tree-list-model-create-model-func)
                                                  :user-data (cffi:make-pointer (glib::put-object nil))
                                                  :user-destroy (cffi:callback glib::free-object-callback)))
           (column-view (make-column-view :model (make-multi-selection :model tree-list-model))))

      (let* ((factory (make-signal-list-item-factory))
             (col (make-column-view-column :title "Test" :factory factory)))

        (flet ((setup (factory item) (declare (ignore factory))
                 (let ((expander (make-tree-expander))
                       (label (make-label :str "")))
                   (setf (tree-expander-child expander) label
                         (list-item-child item) expander)))

               (bind (factory item) (declare (ignore factory))
                 (let* ((row (gobj:coerce (list-item-item item) 'tree-list-row))
                        (uuid (string-object-string (gobj:coerce (tree-list-row-item row) 'string-object)))
                        (expander (gobj:coerce (list-item-child item) 'tree-expander))
                        (label (gobj:coerce (tree-expander-child expander) 'label)))
                   (setf (label-text label) uuid))))

          (connect factory "setup" #'setup)
          (connect factory "bind" #'bind))

        (column-view-append-column column-view col))

      (setf (widget-vexpand-p column-view) t
            (widget-hexpand-p column-view) t
            (window-title window) "Closting"
            (window-child window) column-view)

      (unless (widget-visible-p window)
        (window-present window)))))

```

```common-lisp
arithmetic error FLOATING-POINT-INVALID-OPERATION signalled
   [Condition of type FLOATING-POINT-INVALID-OPERATION]
 0: ((FLET SB-UNIX::RUN-HANDLER :IN SB-UNIX::%INSTALL-HANDLER) 8 #.(SB-SYS:INT-SAP #X66B8E750BD70) #.(SB-SYS:INT-SAP #X66B8E750BC40))
 1: ("foreign function: call_into_lisp_")
 2: ("foreign function: funcall3")
 3: ("foreign function: interrupt_handle_now_handler")
 4: ((LAMBDA (&REST GIR::ARGS-IN) :IN GIR::BUILD-FUNCTION) #.(SB-SYS:INT-SAP #X66B8E00FDB30))
 5: ((LAMBDA (GTK4::APP) :IN GTK-TEST::COM.CLOSTING.MAIN) #<unused argument>)
 6: ((LAMBDA (&REST GTK4::ARGS) :IN GTK4::ATTACH-RESTARTS) #<GIR::OBJECT-INSTANCE {10045A5D53}>)
 7: ((LAMBDA (GIR::CLOSURE RETURN GIR::N-VALUES GIR::PARAMS GIR::HINT GIR::DATA) :IN "/home/soz/quicklisp/dists/quicklisp/software/cl-gobject-introspection-20230618-git/src/signal.lisp") #.(SB-SYS:INT-SAP..
 8: ((LAMBDA (SB-ALIEN::ARGS-POINTER SB-ALIEN::RESULT-POINTER FUNCTION) :IN "/home/soz/quicklisp/dists/quicklisp/software/cl-gobject-introspection-20230618-git/src/signal.lisp") #<unavailable argument> #<..
 9: ("foreign function: funcall_alien_callback")
10: ("foreign function: #x50000DEB")
11: ((LAMBDA (&REST GIR::ARGS-IN) :IN GIR::BUILD-FUNCTION) #.(SB-SYS:INT-SAP #X66B8E0016930) NIL)
12: (GIO:APPLICATION-RUN #<GIR::OBJECT-INSTANCE {1002411A93}> NIL)
13: (SB-INT:SIMPLE-EVAL-IN-LEXENV (GTK-TEST::CLOSTING) #<NULL-LEXENV>)
14: (EVAL (GTK-TEST::CLOSTING))
15: ((LAMBDA NIL :IN SLYNK-MREPL::MREPL-EVAL-1))
```

I also can't figure out how to return a pointer to the child model from the `GtkTreeListModelCreateModelFunc` callback, using 
`(cffi:make-pointer (glib::put-object (make-string-list :strings (list))))` causes a memory fault under SBCL, and simply hangs under CCL.

I appreciate any help, and thank you very much for the great library.

**Comment by bohonghuang** [comment]
> the CFFI callback seems to cause an internal floating point error with SBCL

It's strange that I haven't encountered this situation in my environment, but you can execute the following code before running the GTK application to avoid this problem:

```lisp
(sb-int:set-floating-point-modes :traps nil)
```

**Comment by bohonghuang** [comment]
> I also can't figure out how to return a pointer to the child model from the `GtkTreeListModelCreateModelFunc` callback

The `glib::put-object` and `glib::get-object` functions simply utilize the additional parameter of the callback (i.e., a numerical value of a C pointer) to pass arbitrary Lisp objects. The pointers themselves used in these functions are meaningless. I have made some modifications to the example for your reference:

```lisp
(defstruct product-info
  name cost sale subseries)

(cffi:defcallback tree-list-model-create-model-func :pointer ((item :pointer) (user-data :pointer))
  (destructuring-bind (product-table . build-uuid-string-list)
      (glib::get-object (cffi:pointer-address user-data))
    (if (cffi:null-pointer-p item) (cffi:null-pointer)
        (let* ((uuid (string-object-string (gobj:pointer-object item 'string-object)))
               (product (gethash uuid product-table))
               (string-list (funcall build-uuid-string-list (product-info-subseries product))))
          (tg:cancel-finalization string-list)
          (gobj:object-pointer string-list)))))

(define-application (:name closting :id "com.closting")
  (define-main-window (window (make-application-window :application *application*))
    (let* ((product-table (make-hash-table :test #'equal))
           (build-uuid-string-list (lambda (products)
                                     (loop :for product :in products
                                           :for uuid := (prin1-to-string (uuid:make-v1-uuid))
                                           :do (setf (gethash uuid product-table) product)
                                           :collect uuid :into uuids
                                           :finally (return (gtk:make-string-list :strings uuids)))))
           (root-model (funcall build-uuid-string-list (list (make-product-info :name "iPhone 15"
                                                                                :cost "799$"
                                                                                :sale "899$"
                                                                                :subseries (list (make-product-info
                                                                                                  :name "iPhone 15 Pro"
                                                                                                  :cost "899$"
                                                                                                  :sale "999$")
                                                                                                 (make-product-info
                                                                                                  :name "iPhone 15 Pro Max"
                                                                                                  :cost "1099$"
                                                                                                  :sale "1199$")))
                                                             (make-product-info :name "Infinix S90" :cost "40$" :sale "60$" :subseries nil)
                                                             (make-product-info :name "Casio Calculator" :cost "50$" :sale "80$" :subseries nil))))
           (tree-list-model (make-tree-list-model :root (make-multi-selection :model root-model)
                                                  :passthrough nil
                                                  :autoexpand nil
                                                  :create-func (cffi:callback tree-list-model-create-model-func)
                                                  :user-data (cffi:make-pointer (glib::put-object (cons product-table build-uuid-string-list)))
                                                  :user-destroy (cffi:callback glib::free-object-callback)))
           (column-view (make-column-view :model (make-single-selection :model tree-list-model))))
      (mapc (lambda (name accessor)
              (let* ((factory (make-signal-list-item-factory))
                     (col (make-column-view-column :title name :factory factory)))
                (flet ((setup (factory item) (declare (ignore factory))
                         (let ((expander (make-tree-expander))
                               (label (make-label :str "")))
                           (setf (tree-expander-child expander) label
                                 (list-item-child item) expander)))
                       (bind (factory item) (declare (ignore factory))
                         (let* ((row (gobj:coerce (list-item-item item) 'tree-list-row))
                                (uuid (string-object-string (gobj:coerce (tree-list-row-item row) 'string-object)))
                                (expander (gobj:coerce (list-item-child item) 'tree-expander))
                                (label (gobj:coerce (tree-expander-child expander) 'label)))
                           (when (string= name "Name")
                             (setf (tree-expander-list-row expander) row))
                           (setf (label-text label) (princ-to-string (funcall accessor (gethash uuid product-table)))))))
                  (connect factory "setup" #'setup)
                  (connect factory "bind" #'bind))
                (column-view-append-column column-view col)))
            (list "Name" "Cost" "Sale")
            (list #'product-info-name #'product-info-cost #'product-info-sale))
      (setf (widget-vexpand-p column-view) t
            (widget-hexpand-p column-view) t
            (window-title window) "Closting"
            (window-child window) column-view)
      (unless (widget-visible-p window)
        (window-present window)))))
```

It appears to be working well:

![screenshot](https://github.com/bohonghuang/cl-gtk4/assets/11132087/361cb283-92dd-4528-b734-6d7e2d7834c3)

**Comment by seigakaku** [comment]
Your example worked perfectly under both CCL and SBCL, thank you. 
I still got the floating point error under SBCL, but running `(sb-int:set-floating-point-modes :traps nil)` got rid of it.

I just don't quite understand the purpose of the line `(tg:cancel-finalization string-list)` in the callback, it seems to work the same without it.

**Comment by bohonghuang** [comment]
> I just don't quite understand the purpose of the line `(tg:cancel-finalization string-list)` in the callback, it seems to work the same without it.

In general, `cl-gobject-introspection` adds a finalizer for each GObject, allowing Lisp to modify the reference count of the GObject during garbage collection to correctly free memory. However, the caller of TreeListModelCreateModelFunc takes ownership of the returned object. If Lisp destroys this object when the caller has already done so during garbage collection, memory issues may occur.

**Comment by proehlen** [comment]
I'm getting the subject issue, "FLOATING-POINT-INVALID-OPERATION", with the equivalent of a simple hello world app also under SBCL but I've noticed the following things:

* It seems to trigger when the mouse is moved
* It only happens for me when running the app under Emacs (via SLY) or VSCode (via ALIVE). Running SBCL directly in the terminal doesn't produce the problem.

I've confirmed the workaround provided above, (sb-int:set-floating-point-modes :traps nil) , prevents this error, but I'm leery of broadly suppressing floating point errors for all parts of my application, not just GTK because I'm an inexperienced Lisp dev. Anyway, reporting the above feedback in case it's useful for further debugging.
